### DISCUSSION TITLE
[Problem] How to reset all the data to make Shinobu work again？

### DISCUSSION THREAD

**Discussion by hucurve** [Difegue/LANraragi#1010] [Q&A]
I have been using lanraragi for almost a year now. I really love and appreciate the fact that someone can make software that is so highly customizable and as interactive and friendly as possible.

However, there are still some major bugs that make me very sad. The garbled file names are visually distracting at best, and the occasional missing file is truly maddening.

I found that every time I tried to use the cleanup database option, I thought it would lighten the load on the system. It actually pops up with 200+ files cleaned (I'm up to about 500 total), the corresponding files disappear upon refresh, and despite all attempts to re-read the files, Shinobu ignores them. They are there, but seem to be isolated forever. I don't have a background of computer-related expertise, I'm using docke desktop to mount the image, so I made the following attempts: (which basically failed)
1. update the software version (0.89 to 0.91, only succeeded once)
2. Delete all files I could find in /databse and /content, backup files and thumbnails, and use the reset database option.
3. remount the image and replace the folder with a new one.
4. try using windows desktop version

None of this has worked, and I think the missing file ids are actually still marked by Shinobu.
I was even hoping to re-modify the ids of the files to make them new, but I can't!

i got 520+ files in total about 14.5G.but only 300+ files readed(about 5G)
![image](https://github.com/Difegue/LANraragi/assets/78189608/b493cba2-b2ac-4165-9c07-62163912e21b)
![image](https://github.com/Difegue/LANraragi/assets/78189608/a22c71ed-0110-4e83-ba82-526c721dc9a9)

Docker shows
[LANraragi] [info] Successful login attempt from 172.17.0.1
[Shinobu] [info] Shinobu File Watcher started.
[Shinobu] [info] Content folder is /home/koyomi/lanraragi/content.
[Shinobu] [info] Scanning content folder for changes...
[Shinobu] [info] Found 304 new files.
[Shinobu] [info] 0 files were found on the filemap but not on the filesystem.

lanraragi is a very handy tool and I understand that this fatal bug is very difficult to deal with. 
All I'm hoping for now is to find a way to re-read the missing files (no matter what)
Finally I'd like to thank the author for his hard work!Thank U!

**Comment by Difegue** [comment]
There's no specific data storage for Shinobu alone, all it "remembers" is the filemap in the database and that's wiped when you click "rescan content folder".  

I'm not sure why you'd have 200 files not being picked up unfortunately. File watching support on windows is spotty in that it won't detect newly added files, but doing a rescan/restarting the server normally picks them up.

**Reply by xzero0911** [reply]
I have same issue and research today.
I rebuild all database and not solved issue even rescan or restart WSL. 
But I found if I add small size file that Shinobu can pickup them as well.

For example, DB have 6730 files now
I add 10 new big size(10MB per .jpg) of issue archive to rescan and lost some file others normal archive.
[Shinobu] [info] Shinobu File Watcher started.
[Shinobu] [info] Content folder is /mnt/d/同人.
[Shinobu] [info] Scanning content folder for changes...
[Shinobu] [info] Found 6703 new files.

But if I resize the issue jpg to smaller(3MB per .jpg) and add 10 archive new file. And Shinobu can pickup them as well.
[Shinobu] [info] Shinobu File Watcher started.
[Shinobu] [info] Content folder is /mnt/d/同人.
[Shinobu] [info] Scanning content folder for changes...
[Shinobu] [info] Found 6740 new files.

So, I guess the files too much in one archive or size too big or I/O issue that reach timeout when Shinobu scanning content.
Is it possible add more timeout let Shinobu can finish her job although I/O slow?

**Comment by ilevalser** [comment]
I ran into a similar problem with version 0.91 of wsl 2 for windows, whenever I put new files into the contents folder and rescan it, Shinobu doesn't recognize them correctly, restarting Lanraragi and rebooting the computer doesn't work.
I used "Run as administrator" on Karen.exe and it successfully recognized the zip that it had previously failed to recognize. But strangely, when all the new files are added to the library, the total number of files in the library is more than the number of files in the contents folder.
![Weixin Image_20240705200709](https://github.com/Difegue/LANraragi/assets/145573665/3a1b64f4-a71e-41b9-bf1c-630664e23fee)
![Weixin Image_20240705200752](https://github.com/Difegue/LANraragi/assets/145573665/2b804ffe-6225-47fb-8acc-e5843f1d983e)
 I think some files may have been duplicated, but I don't know which ones because there are too many files.

**Comment by hucurve** [comment]
God, I had this problem again today. I've opened WSL2 on Docker desktop, and 700 files have been unlinked after choosing to clean up the repository, followed by a refreshed, empty repository. Using a backup restore is also completely unchanged. I really hope this problem can be solved as soon as possible. I can only write tag in the file name now, because they will not be lost with the bug😢

**Comment by njjzx** [comment]
I meet the same problem in ver 0.9.31. I have more than 15000 files but LRR only scans 150+
But when I go back to ver 0.9.0, it works correctly.

**Reply by Difegue** [reply]
0.9 to 0.9.31 is a pretty big jump -- it'd be helpful to narrow down which version exactly broke scanning for you but I imagine it's one of the versions that updated the base alpine image.  

There's really no way to improve filewatching on windows as-is except by offloading some of that to the desktop sideapp... (Basically replace Shinobu by a windows threadloop in Karen)  
But then you get into needing to write to the database or do IPC from windows and that sounds pretty dangerous 🤔

**Reply by njjzx** [reply]
I did not use LRR on windows, I use it on my NAS which OS is debian. And LRR run in docker.
Now the version 0.9.0 is OK for me to manage my manga folder and I will first try to get metadata for all manga file. Maybe next month I will have time to narrow down the version to find out which version cause this bug.

**Comment by ilevalser** [comment]
Almost a year later, I wanted to solve this problem again.
Using the base distro of WSL (mine is Ubuntu), I entered the library folder and ran "ls". After the error "ls: reading directory '.': Input/output error" was prompted, some file names could be displayed. This prompted me whether a certain file caused the error.
I found that Linux only supports reading 255-byte file paths. I used a python script to check whether the number of bytes occupied by the file name is greater than 237 (255-18) after deducting the path length (mine is "/mnt/g/DOUJINSHI/"), and as expected, 10 files were found. 
![image](https://github.com/user-attachments/assets/39b4048e-8fcf-4a64-bef3-14358f7bce12)
After moving them out of the folder, the "ls" command can be used normally, and karen can also recognize all files normally.
If you are used to saving files with Japanese titles, please check whether the file name length exceeds the limit. If it does, the scan will stop with an error after the first overlong file path is found, and the performance in LRR is that many files cannot be scanned.
Since this is a root problems of the Linux, you can only modify the file name to be shorter to solve this problem.
